### PR TITLE
[Doc|Train] Add Pytorch ResNet finetuning starter example

### PR DESCRIPTION
**Pull request by woshiyyya**
## Why are these changes needed?

This example aims to provide an example of how to do finetuning on GPUs with TorchTrainer.
```
- Ray Train
   - examples
     - ...
     - Finetuning a Pytorch Image Classifier with Ray AIR
```
Task: Image Classification
Model: Pretrained Pytorch ResNet-50
Dataset: [hymenoptera_data](https://www.kaggle.com/datasets/ajayrana/hymenoptera-data) (2-class, ants and bees)

The workflow is:

- Download hymenoptera_data dataset and create a ray dataset
- Define the training loop (Load a pre-trained model)
- Define the ScalingConfig and RunConfig
- Build a TorchTrainer and run `Trainer.fit()`
- Load the model weights from checkpoint file, and run an evaluation loop in pure PyTorch.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by richardliaw — COMMENTED** [review]
any reason we need to use ray data here?

**Comment by woshiyyya** [comment]
> any reason we need to use ray data here?

Oh I should replace it with `prepare_dataloader`.

**Comment by woshiyyya** [comment]
> any reason we need to use ray data here?

I've replaced ray data with prepare_dataloader().

**Comment by amogkam** [comment]
should this be in train examples or air examples?

**Comment by woshiyyya** [comment]
> should this be in train examples or air examples?

I created a bookmark under the "Ray Train - Examples" section in the document.

Regarding the .ipynb file, I placed it under `ray-air/examples` since I could not find any other notebooks under the `doc/source/train/examples` folder. Is it acceptable to keep it there? Or we should move it?

**Comment by richardliaw** [comment]
I think we should put it in train examples for now.

File: doc/source/ray-air/examples/pytorch_resnet_finetune.ipynb [hunk 1]
```diff
@@ -0,0 +1,765 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Finetuning a Pytorch Image Classifier with Ray AIR\n",
+    "In this example we will finetune a pretrained ResNet model with Ray Train. \n",
+    "\n",
+    "For this example, our network architecture consists of the intermediate layer output of a pretrained ResNet model, which feeds into a randomly initialized linear layer that outputs classification logits for our new task.\n",
```
Comment (angelinalg) [inline doc/source/ray-air/examples/pytorch_resnet_finetune.ipynb:10]:
```suggestion
    "For this example, the network architecture consists of the intermediate layer output of a pre-trained ResNet model, which feeds into a randomly initialized linear layer that outputs classification logits for our new task.\n",
```

File: doc/source/ray-air/examples/pytorch_resnet_finetune.ipynb [hunk 2]
```diff
@@ -0,0 +1,765 @@
...
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {
+    "tags": [
+     "remove-cell"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "# If you want to run full example, please set SMOKE_TEST as False\n",
```
Comment (angelinalg) [inline doc/source/ray-air/examples/pytorch_resnet_finetune.ipynb:34]:
```suggestion
    "# To run full example, set SMOKE_TEST as False\n",
```

File: doc/source/ray-air/examples/pytorch_resnet_finetune.ipynb [hunk 3]
```diff
@@ -0,0 +1,765 @@
...
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Load and preprocess finetuning dataset\n",
+    "This example is adapted from Pytorch's [Fintuning Torchvision Models](https://pytorch.org/tutorials/beginner/finetuning_torchvision_models_tutorial.html) tutorial.\n",
+    "We will use *hymenoptera_data* as the finetuning dataset, which contains two classes (bees and ants) and 397 total images (across training and validation). This is a quite small dataset and we use this only for demenstration purpose. "
```
Comment (angelinalg) [inline doc/source/ray-air/examples/pytorch_resnet_finetune.ipynb:21]:
```suggestion
    "We are using *hymenoptera_data* as the finetuning dataset. It contains two classes (bees and ants) and 397 total images (across training and validation). This small dataset is for demonstration purposes. "
```

File: doc/source/ray-air/examples/pytorch_resnet_finetune.ipynb [hunk 4]
```diff
@@ -0,0 +1,765 @@
...
+   "source": [
+    "# If you want to run full example, please set SMOKE_TEST as False\n",
+    "SMOKE_TEST = True\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The dataset is publicly available [here](https://www.kaggle.com/datasets/ajayrana/hymenoptera-data). Notice that the dataset is structured with directory names as the labels. We can use `torchvision.datasets.ImageFolder()` to load the images and their corresponding labels."
```
Comment (angelinalg) [inline doc/source/ray-air/examples/pytorch_resnet_finetune.ipynb:42]:
```suggestion
    "The dataset is publicly available [here](https://www.kaggle.com/datasets/ajayrana/hymenoptera-data). Note that it is structured with directory names as the labels. Use `torchvision.datasets.ImageFolder()` to load the images and their corresponding labels."
```

**Comment by angelinalg** [comment]
I'd like to do more copy editing for you but the JSON file was a bit difficult to use. Is there an easier way for me to provide suggestions on the copy?

File: doc/source/train/examples/pytorch/BUILD [hunk 5]
```diff
@@ -0,0 +1,5 @@
+filegroup(
+    name = "train_pytorch_examples",
+    srcs = glob(["*.ipynb"]),
+    visibility = ["//doc:__subpackages__"]
+)
```
Comment (justinvyu) [inline doc/source/train/examples/pytorch/BUILD:5]: newline

File: doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb [hunk 6]
```diff
@@ -0,0 +1,776 @@
...
+    "    model.fc = nn.Linear(num_features, 2)\n",
+    "\n",
+    "    # Ensure all params get updated during fintuning\n",
+    "    for param in model.parameters():\n",
+    "        param.requires_grad = True\n",
+    "    return model\n",
+    "\n",
+    "\n",
+    "# Option 2: Initialize model with an AIR checkpoint\n",
+    "CHECKPOINT_URI = \"s3://air-example-data/finetune-resnet-checkpoint/TorchTrainer_4f69f_00000_0_2023-02-14_14-04-09/checkpoint_000000/\"\n",
```
Comment (justinvyu) [inline doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb:189]: Add a comment for the user to replace this with their own s3 uri.

File: doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb [hunk 7]
```diff
@@ -0,0 +1,776 @@
...
+    "\n",
+    "**2. Prepare your model**:\n",
+    "- {meth}`train.torch.prepare_model() <ray.train.torch.prepare_model>` prepares the model for distributed training. Under the hood, it converts your torch model to `DistributedDataParallel` model, which synchronize its weights across all workers.\n",
+    "\n",
+    "**3. Report metrics and checkpoint**:\n",
+    "- {meth}`session.report() <ray.air.session.report>` will report metrics and checkpoints to Ray AIR.\n",
+    "- Saving checkpoints through {meth}`session.report(metrics, checkpoint=...) <ray.air.session.report>` will automatically [upload checkpoints to cloud storage](tune-cloud-checkpointing) (if configured), and allow you to easily enable Ray AIR worker fault tolerance in the future.\n",
+    "- The best checkpoints are saved according to the specified `checkpoint_score_attribute` in {class}`CheckpointConfig <ray.air.config.CheckpointConfig>`. Here we only save the best model with highest validation accuracy."
```
Comment (justinvyu) [inline doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb:218]: We can remove this sentence, we don't do this in the example.

File: doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb [hunk 8]
```diff
@@ -0,0 +1,776 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from ray.train.torch import TorchTrainer, TorchCheckpoint\n",
+    "from ray.air.config import ScalingConfig, RunConfig, CheckpointConfig\n",
+    "from ray.tune.syncer import SyncConfig\n",
```
Comment (justinvyu) [inline doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb:348]: Can remove this

File: doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb [hunk 9]
```diff
@@ -0,0 +1,776 @@
...
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "import os\n",
+    "\n",
+    "checkpoint_folder = result.checkpoint.uri.replace(\"file://\", \"\")\n",
+    "print(checkpoint_folder)\n",
+    "os.listdir(checkpoint_folder)\n"
```
Comment (justinvyu) [inline doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb:665]: Can we do `TorchCheckpoint.from_uri(...)` to handle the `file://`? Also, let's not show the contents of the checkpoint folder - not too useful to see.

File: doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb [hunk 10]
```diff
@@ -0,0 +1,776 @@
...
+      "  warnings.warn(\n",
+      "/home/ray/anaconda3/lib/python3.9/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=ResNet50_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet50_Weights.DEFAULT` to get the most up-to-date weights.\n",
+      "  warnings.warn(msg)\n"
+     ]
+    }
+   ],
+   "source": [
+    "checkpoint = TorchCheckpoint.from_directory(checkpoint_folder)\n",
+    "model = checkpoint.get_model(initialize_model())\n",
+    "device = torch.device(\"cuda\")\n"
```
Comment (justinvyu) [inline doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb:695]: We already have an `initialize_method_from_checkpoint` method defined earlier. Let's use that and pass in the `file://` URI.

File: doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb [hunk 11]
```diff
@@ -0,0 +1,776 @@
...
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Load and preprocess finetuning dataset\n",
+    "This example is adapted from Pytorch's [Fintuning Torchvision Models](https://pytorch.org/tutorials/beginner/finetuning_torchvision_models_tutorial.html) tutorial.\n",
+    "We will use *hymenoptera_data* as the finetuning dataset, which contains two classes (bees and ants) and 397 total images (across training and validation). This is a quite small dataset and we use this only for demenstration purpose. "
```
Comment (justinvyu) [inline doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb:21]:
```suggestion
    "We will use *hymenoptera_data* as the finetuning dataset, which contains two classes (bees and ants) and 397 total images (across training and validation). This is a quite small dataset and used only for demonstration purposes. "
```

File: doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb [hunk 12]
```diff
@@ -0,0 +1,776 @@
...
+    "\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Load and preprocess finetuning dataset\n",
+    "This example is adapted from Pytorch's [Fintuning Torchvision Models](https://pytorch.org/tutorials/beginner/finetuning_torchvision_models_tutorial.html) tutorial.\n",
```
Comment (justinvyu) [inline doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb:20]:
```suggestion
    "This example is adapted from Pytorch's [Finetuning Torchvision Models](https://pytorch.org/tutorials/beginner/finetuning_torchvision_models_tutorial.html) tutorial.\n",
```

File: doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb [hunk 13]
```diff
@@ -0,0 +1,776 @@
...
+    "# Option 1: Initialize model with pretrained weights\n",
+    "def initialize_model():\n",
+    "    # Load pretrained model params\n",
+    "    model = models.resnet50(pretrained=True)\n",
+    "\n",
+    "    # Replace the original classifier with a new Linear layer\n",
+    "    num_features = model.fc.in_features\n",
+    "    model.fc = nn.Linear(num_features, 2)\n",
+    "\n",
+    "    # Ensure all params get updated during fintuning\n",
```
Comment (justinvyu) [inline doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb:182]:
```suggestion
    "    # Ensure all params get updated during finetuning\n",
```

**Review by justinvyu — APPROVED** [review]
I think it's pretty much ready - have some minor suggestions:

One question - should we showcase `TorchTrainer(resume_from_checkpoint)` as the way to take in the AIR checkpoint in this example?

File: doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb (hunk 9 above)
Comment (woshiyyya) [reply doc/source/train/examples/pytorch/pytorch_resnet_finetune.ipynb:665]: Ok

**Review by richardliaw — APPROVED** [review]
clean and concise!

**Pull request merged**